kola/tests/update/update.go

```diff
@@ -0,0 +1,172 @@
+// Copyright 2018 CoreOS, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package update
+
+import (
+	"bufio"
+	"fmt"
+	"os"
+	"strings"
+	"time"
+
+	"github.com/coreos/go-omaha/omaha"
+
+	"github.com/coreos/mantle/kola"
+	"github.com/coreos/mantle/kola/cluster"
+	"github.com/coreos/mantle/kola/register"
+	tutil "github.com/coreos/mantle/kola/tests/util"
+	"github.com/coreos/mantle/platform"
+	"github.com/coreos/mantle/platform/local"
+	"github.com/coreos/mantle/util"
+)
+
+func init() {
+	register.Register(&register.Test{
+		Name:        "coreos.update.payload",
+		Run:         payload,
+		ClusterSize: 1,
+		NativeFuncs: map[string]func() error{
+			"Omaha": Serve,
+		},
+	})
+}
+
+func Serve() error {
+	omahaserver, err := omaha.NewTrivialServer(":34567")
+	if err != nil {
+		return fmt.Errorf("creating trivial omaha server: %v\n", err)
+	}
+
+	omahawrapper := local.OmahaWrapper{TrivialServer: omahaserver}
+
+	if err = omahawrapper.AddPackage("/updates/update.gz", "update.gz"); err != nil {
+		return fmt.Errorf("bad payload: %v", err)
+	}
+
+	return omahawrapper.Serve()
+}
+
+func payload(c cluster.TestCluster) {
+	addr := ConfigureOmahaServer(c, c.Machines()[0])
+
+	// create the actual test machine, the machine
+	// that is created by the test registration is
+	// used to host the omaha server
+	m, err := c.NewMachine(nil)
+	if err != nil {
+		c.Fatalf("creating test machine: %v", err)
+	}
+
+	ConfigureMachineForUpdate(c, m, addr)
```

Comment:
Is there a reason the machine isn't being configured with Ignition?

That'll make it easier to not race with update-engine for writing files (since the systemd unit that bindmounts over the pubkey can have a `Before=update-engine.service`).

I also think a CLC is more readable than the procedural ssh bits in this function.

Reply:
My initial thought for this was to leave the UserData completely up to the user so different tests could configure machines however they wanted and then these functions would do post-configuration.

Now that we're shipping an updated Ignition vendor I'll look into adding some new methods into `platform/conf` for combining configs and provide a function in this module that takes a custom userdata and adds in the requisite pieces for the update configuration.

Reply:
After thinking about it again, it actually makes sense to not use Ignition since we may want to test updates on versions which predate ignition.

I think a comment to that effect on the function would be worth having.